File: LeetCode/0390. Elimination Game/solution.py
"""
3377 / 3377 test cases passed.
Runtime: 44 ms
Memory Usage: 15 MB
"""
class Solution:
    def lastRemaining(self, n: int) -> int:
        head, step, left = 1, 1, True
        while n > 1:
            if left or n & 1 == 1:
                head += step
            step <<= 1
            n >>= 1
            left = not left
        return head
    

File: LeetCode/0390. Elimination Game/solution.cpp
/**
 * 3377 / 3377 test cases passed.
 * Runtime: 0 ms
 * Memory Usage: 5.8 MB 
 */
class Solution {
public:
    int lastRemaining(int n) {
        int head = 1;
        int step = 1;
        bool left = true;
        while (n > 1) {
            if (left || n & 1) {
                head += step;
            }
            step <<= 1;
            left = !left;
            n >>= 1;
        }
        return head;
    }
};
